### ISSUE TITLE
Service object purge is limited to 1000 services

### ISSUE THREAD

**Issue by wp-perc** [Icinga/icingaweb2-module-director#2339]
## Expected Behavior
When a Sync Rule has to purge 100+ objects, it should do it.

## Current Behavior
If a Sync Rule has to purge more than 1000 objects, it only removes the first 1000.

## Possible Solution
Let this limit be configurable, or removable with a FORCE option

## Steps to Reproduce (for bugs)
Create a Import Source with 10000 rows
Create a Rule that uses that Source to create Host Objects, and set the Purge option inside the Rule to Yes
Change the Import Source to drop the imported lines to 0
Run the Sync Rule: it will purge only 1000 objects

Reason:
ImportRunBasedPurgeStrategy.php, line number 75:
foreach (array_chunk($result, **1000**) as $keys) {

**Comment by lazyfrosch** [comment]
`array_chunk()` should not be the cause of the problem here.

`$result` contains the imported rows with object names that where in runA, but not in runB.

This result is then split into chunks of 1000, meaning `$keys` can be up to 1000 entries for object name. For those chunks of objects, the full row with all columns is retrieved (to limit query size) and the final object name is built (for services e.g.).

Then an array of names is returned that disappeared between runA and runB.

---

Could you please elaborate what kind of import causes you problems?

I remember that Active Directory has a basic limit of 1000 results, and heard about other problems with the current pagination of LDAP queries in Icinga Web 2 a while ago.

**Comment by wp-perc** [comment]
The import source is not important. What is important here is the Sync Rule.
However, in my scenario, the import source is a Director Object, not Active Directory, and the Import Source reads correctly all data: I checked on the History, and there was definitely more than 1k rows.

About my scenario...
I had to remove about 17k (yes, 17 thousands) Service Objects from Director because of a drastic change of naming convention.
To ensure everything is alright, I decided to proceed in two steps:
 - step1: removal of the old service objects
 - step2: creation of the new service objects

To perform step1 quickly, I decided to take advantage of the PURGE feature of Sync Rules.
Here what I did:
1. Create a new Import Source of type Director Object containing all the SO (Service Objects) I need to remove
2. Run Import
3. Create a Sync Rule with PURGE that creates these SO in merge mode
4. Trigger sync (this way, a link between the Sync Rule and the SO I want to remove is created)
4. Edit the Import Source to drop all lines (with a Modifier)
5. Run Import again: this way, all SO are then ready to be purged by the Sync Rule
6. Check for changes on the Sync Rule: it showed me 17k objects will be dropped
7. Trigger Sync: I was expecting for all the 17k objects to drop, **but only 1000 have been removed**
8. Trying with another Trigger Sync, but nothing: the Sync Rule says it has done all that he had to, but the remaining 16k objects are still alive and kicking inside Director

To complete the step 1, I had to:
1. Edit the Import Source to include the yet-to-be-removed-SOs
2. Run Import
3. Trigger Sync: nothing happens, but the Sync Rule links to the objects I want to remove
4. Edit the Import Source to drop all lines
5. Run Import
6. Trigger Sync: another batch of 1k SO is removed
7. Go back to 1.

After some time, I looked into the code, found line line number 75 on ImportRunBasedPurgeStrategy.php, changed the `foreach (array_chunk($result, 1000) as $keys) {` to `foreach (array_chunk($result, 10000) as $keys) {` and ALL SOs has been removed in one run. So, I'm sorry but the issue IS related to `array_chunk()`, but maybe is the whole algorithm that is not right...

**Comment by Thomas-Gelf** [comment]
@wp-perc, @lazyfrosch: you're both right. There IS a bug, it's in that loop - but array_chunk is not to blame here. I introduced this bug when introducing chunking, as there is the same `$result` variable being used for two different tasks. This was wrong from the very beginning, but became a problem only after chunking has been added to the mix. Right now, it is emptied every time you enter the loop. Depending on your PHP version, this could also lead to unpredictable results when not exceeding 1000 elements.

Will be fixed.
